### DISCUSSION TITLE
Connecting to a MYSQL database

### DISCUSSION THREAD

**Discussion by RickRyan26** [vercel/next.js#19162] [Help] [answered]
Hello!

I followed the entire Vercel Next.js Mysql Tutorial below and am getting the following error.

https://vercel.com/guides/deploying-next-and-mysql-with-vercel

Note {protocal} and {host} are both console.loging as "undefined".
The deployed page shows "502: BAD_GATEWAY Code: NO_RESPONSE_FROM_FUNCTION" with no errors in its logs.
Any thoughts? Thank you!

![image](https://user-images.githubusercontent.com/72894373/99133044-3dc86a00-25cd-11eb-8d3b-26555c63fb51.png)

I'm really struggling to figure this one out =(

**Comment by leerob** [comment]
If both `protocol` and `host` are `undefined`, then you'll get a 502 because you're trying to connect to an invalid URL. Your URL would look something like this:

```
undefined://undefined/api/projects...
```

Could you share more of your code?

**Comment by RickRyan26** [comment]
Hi Leerob, thanks for your response!

Makes senses to me. After studying further I guess what i'm really not understanding is how req gets defined before getServerSideProps, which leaves me confused as to why 'x-forwarded-proto' is undefined in it. Just about all of the code im using can be found in the tutorial here https://vercel.com/guides/deploying-next-and-mysql-with-vercel but to save you the time here it is. I definitely feel like im missing something simple.

**lib/db.js**
```
const db = mysql({
    config: {
      host: process.env.mysql_host,
      database: process.env.mysql_database,
      user: process.env.mysql_user,
      password: process.env.mysql_password,
    },
  })
  
  exports.query = async (query) => {
    try {
      const results = await db.query(query)
      await db.end()
      return results
    } catch (error) {
      return { error }
    }
  }
```

The above config is set with Vercel CLI secrets and is accessing a GoDaddy MySQL database. I have the database whitelisting 0.0.0.0 and %.%.%.% since vercels IPs are dynamic.

**vercel.json**
```
{
  "env": {
    "mysql_host": "@mysql_host",
    "mysql_user": "@mysql_user",
    "mysql_password": "@mysql_password",
    "mysql_database": "@mysql_database"
  }
}
```

**api/profiles/profile.js**
```
const db = require('../../lib/db')
const escape = require('sql-template-strings')

module.exports = async (req, res) => {
  const [profile] = await db.query(escape`
    SELECT *
    FROM invoicesdev
    WHERE id = ${req.query.id}
  `)
  res.status(200).json({ profile })
}
```

**pages/profile.js**
```
import Link from 'next/link'

function ProfilePage({ profile }) {
  return (
    <>
      <div>
        <h1>{profile.gate}</h1>
        <p>{profile.fence}</p>
        <p>{profile.auto}</p>
        <Link href="/">
          <a>← Back to profiles</a>
        </Link>
      </div>
    </>
  )
}

export async function getServerSideProps({ req, query }) {
  const protocol = req.headers['x-forwarded-proto']
  const host = req.headers['x-forwarded-host']

  const res = await fetch(`${protocol}://${host}/api/profiles/${query.id}`)
  const data = await res.json()

  return { props: data }
}

export default ProfilePage
```

**api/profiles/index.js**
```
const db = require('../../lib/db')
const escape = require('sql-template-strings')

module.exports = async (req, res) => {
  let page = parseInt(req.query.page) || 1
  const limit = parseInt(req.query.limit) || 9
  if (page < 1) page = 1
  const profiles = await db.query(escape`
      SELECT *
      FROM invoicesdev
      ORDER BY id
      LIMIT ${(page - 1) * limit}, ${limit}
    `)
  const count = await db.query(escape`
      SELECT COUNT(*)
      AS profilesCount
      FROM invoicesdev
    `)
  const { profilesCount } = count[0]
  const pageCount = Math.ceil(profilesCount / limit)
  res.status(200).json({ profiles, pageCount, page })
}
```

**pages/index.js**
```
import Link from 'next/link'

function HomePage({ profiles, page, pageCount }) {
  return (
    <>
      <ul>
        {profiles.map((p) => (
          <li className="profile" key={p.id}>
            <Link href={`/profile?id=${p.id}`}>
              <a>
                <img src={p.gate} />
                <span>{p.fence}</span>
              </a>
            </Link>
          </li>
        ))}
      </ul>
      <nav>
        {page > 1 && (
          <Link href={`/?page=${page - 1}&limit=9`}>
            <a>Previous</a>
          </Link>
        )}
        {page < pageCount && (
          <Link href={`/?page=${page + 1}&limit=9`}>
            <a className="next">Next!</a>
          </Link>
        )}
      </nav>
    </>
  )
}

export async function getServerSideProps({ req, query }) {
  console.log(req);
  const protocol = req.headers['x-forwarded-proto']
  const host = req.headers['x-forwarded-host']
  const page = query.page || 1
  const limit = query.limit || 9

  const res = await fetch(
    `${protocol}://${host}/api/profiles?page=${page}&limit=${limit}`
  )
  const data = await res.json()

  return { props: data }
}

export default HomePage
```

Thanks again! Any thoughts help

**Reply by leerob** [reply]
Here's [more information on `req`](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering), which uses [this](https://nodejs.org/api/http.html#http_class_http_incomingmessage) under the hood.

It looks like those headers aren't there when I tested locally. If you `console.log(req.headers)` you should see the options you have. You're basically trying to determine if you're local or deployed, since local you're on `http://localhost:3000` and deployed you're `https://your-app.com` for example. 

`headers.referer` might work 👍

**Comment by RickRyan26** [comment]
That got me a little further! Thanks.

Now im ending up with this error:

![image](https://user-images.githubusercontent.com/72894373/99199436-f0c2d000-2753-11eb-92f4-518117f12d03.png)

Which is interesting because res does not start with "<"; it's the following:

```
Response {
  size: 0,
  timeout: 0,
  [Symbol(Body internals)]: {
    body: Gunzip {
      _writeState: [Uint32Array],
      _readableState: [ReadableState],
      _events: [Object: null prototype],
      _eventsCount: 5,
      _maxListeners: undefined,
      _writableState: [WritableState],
      allowHalfOpen: true,
      bytesWritten: 0,
      _handle: [Zlib],
      _outBuffer: <Buffer 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 55 20 d7 23 06 00 00 00 00 b0 8d 2d 06 00 00 00 00 65 13 db d2 f5 b0 e4 dc 8d 44 58 5c 84 12 47 b8 e6 1f ... 16334 more bytes>,
      _outOffset: 0,
      _chunkSize: 16384,
      _defaultFlushFlag: 2,
      _finishFlushFlag: 2,
      _defaultFullFlushFlag: 3,
      _info: undefined,
      _maxOutputLength: 4294967295,
      _level: -1,
      _strategy: 0,
      [Symbol(kCapture)]: false,
      [Symbol(kTransformState)]: [Object],
      [Symbol(kError)]: null
    },
    disturbed: false,
    error: null
  },
  [Symbol(Response internals)]: {
    url: 'http://localhost:3000/api/profiles?page=1&limit=9',
    status: 404,
    statusText: 'Not Found',
    headers: Headers { [Symbol(map)]: [Object: null prototype] },
    counter: 0
  }
}
```
So now my guess is somethings wrong in api/profiles/index.js but idk what...

**Reply by leerob** [reply]
I’ll give you a hint.. look at the status code and status text in what you just sent 😁

**Reply by RickRyan26** [reply]
Haha I see that it's a 404, which explains the "<" but that's definitely the path to /api/profiles/index.js , I've quadruple checked it. I assume that's what res.status(200).json() is supposed to prevent too... Im so confused and I just want this to work so bad 😂😭

**Reply by leerob** [reply]
A few debugging ideas:

1. Are you sure your server is running at localhost:3000?
2. Can you hit the API route directly in the browser since it's a GET? Does it 404 there?

**Reply by RickRyan26** [reply]
Yes and yes. Please see below =(

![image](https://user-images.githubusercontent.com/72894373/99205193-997d2980-276c-11eb-9677-8ae66845f865.png)

**Reply by leerob** [reply]
The `api` folder is supposed to be inside `pages`. So `pages/api/profiles/index.js` 👍

**Reply by leerob** [reply]
Otherwise, you'd need to run `vercel dev` to start the serverless functions inside the `/api` folder. When they're nested underneath `pages`, you can just run `yarn dev` to start Next.js and it handles both.

**Reply by RickRyan26** [reply]
Moving the api folder into pages folder goes against the Vercel tutorial but I tried it anyways since it makes sense to me and got the below. It switched us from a 404 to a 500. 

![image](https://user-images.githubusercontent.com/72894373/99206035-35a83000-276f-11eb-80fb-018407730594.png)

Also I've been running ```npm run dev``` is that an issue? Seems to be doing the same as ```yarn dev```

```vercel dev``` returns me the following. Not sure what it means.

![image](https://user-images.githubusercontent.com/72894373/99206528-76ed0f80-2770-11eb-8f37-8b5243fe0b06.png)

Thanks again! I'm losing my mind over here spending days on this haha

**Reply by leerob** [reply]
You don't need to use `vercel dev`, I was just explaining why. `yarn dev` or `npm run dev` are exactly the same thing, just different package managers. The fact that it's a 500 instead of a 404 means that it's correctly registering the route, so you're past that issue. 

I think you have too many pieces in place to effectively debug this. Start from square one.

1. Make a brand new API route at `pages/api/hello.js`
2. Ensure that you can correctly hit that API with a 200 and you get data back
3. Then, add the code from your previous API that isn't working
4. Add console logs (or other debugging) to figure out what isn't right with your code

The guide needs updated as it's obviously not correct anymore (it's a few years ago) 😄 I'd recommend reading through the docs first so you have a better understanding of how things work: https://nextjs.org/docs/api-routes/introduction

**Reply by RickRyan26** [reply]
Great advice! I'm running out of time working on this today, but I'll get right back on this tomorrow and start your 4 step plan! Bummer I didn't realize the tutorial was outdated.

To keep you updated tho, before I saw your latest message I did some investing into ```vercel dev```, went through it's install process, created an /.env file with my MySQL credentials, and then ran ```vercel dev```.

First thing I noticed was our req variable now contains our desired header props:

```
const protocol = req.headers['x-forwarded-proto']
const host = req.headers['x-forwarded-host']
```

Then I noticed the following:

![image](https://user-images.githubusercontent.com/72894373/99208605-cd108180-2775-11eb-9c64-2c9a032c0f81.png)

Turns out if keep the api directory inside of the pages directory we get the same 500 error as before.
But if we keep the api directory at the root instead we now get the below styled 502 (which matches the 502 page on the deployed version).
Feels like progress.

![image](https://user-images.githubusercontent.com/72894373/99208916-af8fe780-2776-11eb-99ab-3247ae02e255.png)

**Reply by leerob** [reply]
That's because `vercel dev` isn't meant to be used with Next.js. The functionality of `vercel dev` (running serverless functions) is built directly into Next.js, which is why you can do `yarn dev`. That's a good point that the headers must be a bit different between the two.

**Reply by RickRyan26** [reply]
Thank you so much Leerob! I couldn't give it up, finally got this working, and I am stoked about it. You're right the tutorial had quite a bit wrong with it but the final problem was my fault. I found an error log sent exclusively by the new 502 page saying I forgot to import the serverless MySQL module on the db.js page. I feel ridiculous but I really learned a lot from this so whatever.

Thanks again, you made my day

**Reply by leerob** [reply]
Glad it's working! We'll get the guide updated 💯

**Reply by RickRyan26** [reply]
Hah I had no idea you worked at Vercel this whole time! And I just found your website. You have a ton of helpful stuff on there! I will definitely be using it

**Reply by cyphire** [reply]
I am new to next.js and vercel.  Site i'm building is awesome, but I am struggling with all the complexity and difficulty of using mysql with next.js.  The documentation is good, but the examples, like most complex sites require the user to sift through all sorts of information - nothing ever just posts an example which works.  Similar to gatsby, I think that the vast majority of decent programmers eventually give it up.

Is there any chance you could post your entire example as fixed Rick? I've been working on it for 2 days and have looked at the so many examples of issues including connection issues, server vs dev, etc. It's been a tough slog!

**Reply by leerob** [reply]
Great news, the guide has been updated! See my comment below: https://github.com/vercel/next.js/discussions/19162#discussioncomment-137742

**Comment by leerob** [comment, marked as answer]
The Next.js + MySQL guide has been updated, including a brand new demo 🎉 

Guide: https://vercel.com/guides/deploying-next-and-mysql-with-vercel
Demo: https://next-mysql.vercel.app/

**Reply by gabrielparanhosdev** [reply]
I'm getting the following error:

Unhandled Runtime Error
TypeError: entries.map is not a function

Source
components/entries/index.tsx (7:17) @ Entries

   5 | return (
   6 |   <div>
>  7 |     {entries.map((e) => (
     |             ^
   8 |       <div key={e.id} className="py-2">
   9 |         <Entry id={e.id} title={e.title} content={e.content} />
  10 |       </div>

**Comment by cyphire** [comment]
Hey Lee - Are you sure that it is correct?  I sent a feedback about it.

I think most of the content is missing, doesn't tell the user to do a "npm
i" after the npx, doesn't seem to do the migration, etc.  Could you let me
know if it's me or the content?

Best,

Brian.
_________________________________________
Brian B. Canin    brian.canin@gmail.com

858-945-5601 (cell)

**Reply by leerob** [reply]
Yes, this is correct. Did you see this note?

> You should see a success message if the migration ran successfully. If you got an error, please double check your credentials.

Are you sure your credentials are set up correctly?

**Reply by leerob** [reply]
We can try and make this more clear 👍

**Reply by cyphire** [reply]
I don't know why this just came up...

I had a lot of errors figuring things out, let me give whoever is struggling some feedback... 

First, if you are deploying to https:// your-site.com you should test with https://localhost:3000 not http://localhost:3000.  Lots of articles on how to do this but look for this in your package.json:      "dev": "node server.js -p $PORT", instead of just running the default server.  Make sure your server.js is dishing up https (make a cert) instead of http.

Second prob i ran into:  I was wracking my brain with the secrets, etc.  But i do NOT use vercel any more, i do the same with a digital ocean app.  I created a digital ocean MySql server as well and that does NOT use the default port for MySql.  By adding 'MYSQL_PORT': 25060, to the environment, all was well.

I was UNABLE to get getStaticProps and getPaths (sic) working because of this.  I was never getting any MySQL back when getting the pages under local dev, because they render the server stuff each time in the dev mode.  Since my SQL was not being served with the port, I thought my pages were just not rendering.  Take a look at your port info.

The > error you are getting is just saying that you are NOT doing a good fetch.  You are NOT getting data back from your fetch.  Look at that.  Again, you are not doing a successful fetch.

I hope this helps, if you are still struggling send more info.

**Reply by unimprobable** [reply]
@cyphire For your Digital Ocean database, did you add the port to the MYSQL_HOST var in .env.local? Or did you add a new variable line called "MYSQL_PORT"? I tried both and still haven't been able to complete the migrate script. Thanks!

**Comment by cyphire** [comment]
Oh so sorry for late reply, just noticed this... Let me see if I can still help....

Here is my env.config.js file:

```
const config = {
    DATABASE_HOST: process.env.DATABASE_HOST,
    DATABASE_PORT: process.env.DATABASE_PORT,
    DATABASE_NAME: process.env.DATABASE_NAME,
    DATABASE_USERNAME: process.env.DATABASE_USERNAME,
    DATABASE_PASSWORD: process.env.DATABASE_PASSWORD,
    __DEV__: process.env.NODE_ENV === "development",
    IMAGE_DIRECTORY: process.env.IMAGE_DIRECTORY
};

module.exports = config;
```
Here is my .env file: 

note: the port on Digitalocean is 
25060 not the usual default of 3306!!!!  Maybe this is your problem?

```
NODE_ENV="development"
SERVER_PORT=3000
DATABASE_HOST=digitaloceanxxxxxxxxxxxxxxxxxxxxx
DATABASE_PORT=25060
DATABASE_NAME=xxxxx
DATABASE_USERNAME=xxxxxx
DATABASE_PASSWORD=xxxxxx
IMAGE_DIRECTORY=\usr\nodeserver\do_express\images

```

here is my server.js file :

```
require("dotenv").config();

const express = require("express");
const bodyParser = require("body-parser");
var cors = require('cors');
const app = express();

var fs = require('fs')

// I DONT THINK I USE THESE:::::
// parse requests of content-type: application/json
app.use(bodyParser.json());
// parse requests of content-type: application/x-www-form-urlencoded
app.use(bodyParser.urlencoded({ extended: true }));
// END OF NOT USED.... (I think)


// Then use it before your routes are set up:
app.use(cors());

// simple route
app.get("/", (req, res) => {
  res.json({ message: "Welcome to xxxxxxx  application!" });
});

require("./app/routes/zzzzzzzz.routes.js")(app);
// xxxxxxx:
require("./app/routes/xxxxxxx.routes.js")(app);
require("./app/routes/yyyyyyyy.routes.js")(app);


 app.listen(3000, () => {
  console.log("Server is running http on port 3000 (1).");
});



```
I'll get right back to you if you respond again...

**Reply by unimprobable** [reply]
@cyphire Thanks for the help. I am still stuck on [Step 2 in the tutorial](https://vercel.com/guides/deploying-next-and-mysql-with-vercel), "npm run migrate".

I have tried adding MYSQL_PORT=25060 to both the .env.local file and migrate-db.js script.

How did you run the migrate-db.js script with Digital Ocean?

**Reply by unimprobable** [reply]
I got the migrate-db.js script to work by adding the port like you mentioned. I think I was missing a comma or quotes. Here's my file for anyone else trying to use Digital Ocean:

migrate-db.js

```
const path = require('path')
const envPath = path.resolve(process.cwd(), '.env.local')

console.log({ envPath })

require('dotenv').config({ path: envPath })

const mysql = require('serverless-mysql')

const db = mysql({
  config: {
    host: process.env.MYSQL_HOST,
    database: process.env.MYSQL_DATABASE,
    user: process.env.MYSQL_USERNAME,
    password: process.env.MYSQL_PASSWORD,
    port: process.env.MYSQL_PORT,
  },
})

async function query(q) {
  try {
    const results = await db.query(q)
    await db.end()
    return results
  } catch (e) {
    throw Error(e.message)
  }
}

// Create "entries" table if doesn't exist
async function migrate() {
  try {
    await query(`
    CREATE TABLE IF NOT EXISTS entries (
      id INT AUTO_INCREMENT PRIMARY KEY,
      title TEXT NOT NULL,
      content TEXT NOT NULL,
      created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
      updated_at 
        TIMESTAMP 
        NOT NULL 
        DEFAULT CURRENT_TIMESTAMP 
        ON UPDATE CURRENT_TIMESTAMP
    )
    `)
    console.log('migration ran successfully')
  } catch (e) {
    console.error('could not run migration, double check your credentials.')
    process.exit(1)
  }
}

migrate().then(() => process.exit())
```

.env.local 

```
MYSQL_HOST='insertYourDODBurl.com'
MYSQL_DATABASE= 'insertYourDatabaseName'
MYSQL_USERNAME= 'insertYourDODBuserName'
MYSQL_PASSWORD='insertYourDODBpassword'
MYSQL_PORT='25060'
```

**Reply by leerob** [reply]
Looks like the only difference is the port change. Would you mind updating the example with these changes? 🙏 

https://github.com/vercel/next.js/tree/canary/examples/with-mysql

**Reply by unimprobable** [reply]
Hi @leerob Sure thing, I just made [PR #22395](https://github.com/vercel/next.js/pull/22395) . Let me know if you need me to change anything on that, I think I did it right.
